record/program/src/processor.rs

```diff
@@ -0,0 +1,118 @@
+//! Program state processor
+
+use {
+    crate::{
+        error::RecordError,
+        instruction::RecordInstruction,
+        state::{AccountData, Data},
+    },
+    borsh::{BorshDeserialize, BorshSerialize},
+    solana_program::{
+        account_info::{next_account_info, AccountInfo},
+        entrypoint::ProgramResult,
+        msg,
+        program_error::ProgramError,
+        program_pack::IsInitialized,
+        pubkey::Pubkey,
+        rent::Rent,
+        sysvar::Sysvar,
+    },
+};
+
+/// Instruction processor
+pub fn process_instruction(
+    _program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    input: &[u8],
+) -> ProgramResult {
+    let instruction = RecordInstruction::try_from_slice(input)?;
+    let account_info_iter = &mut accounts.iter();
+
+    match instruction {
+        RecordInstruction::Initialize => {
+            msg!("RecordInstruction::Initialize");
+
+            let data_info = next_account_info(account_info_iter)?;
+            let owner_info = next_account_info(account_info_iter)?;
+            let rent_sysvar_info = next_account_info(account_info_iter)?;
+            let rent = &Rent::from_account_info(rent_sysvar_info)?;
+
+            let mut account_data = AccountData::try_from_slice(*data_info.data.borrow())?;
+            if account_data.is_initialized() {
+                return Err(ProgramError::AccountAlreadyInitialized);
```

Comment:
I've found it pretty useful to add `msg!()` statements for all the error cases, so from the transaction log it's more obvious what went wrong without having to drudge up the program source code

Reply:
Done in 0d65c8d